### ISSUE TITLE
Native usage on Debian testing? (nodejs segfault)

### ISSUE THREAD

**Issue by andres-erbsen** [0xADE1A1DE/CryptOpt#118]
Hi,

I am trying to run CryptOpt on natively on Debian testing because reasons. I'd like to know whether you think this has a chance of working. The current state is that I think I have managed to get both AssemblyLine and CryptOpt to build and link. `asmline` runs, but running CryptOpt exits with a segmentation fault:

Last lines of ./CryptOpt output with DEBUG=1 --verbose:
```
executing cmd gcc -march=native -mtune=native -O3 -fPIC -shared -o /tmp/CryptOpt.cache/uczayywaxs/libcheckfunctions-s1678729657975-bfiat-p4149251.so /home/andreser/CryptOpt/dist/data/fiat-bridge/.cache/85ea43d3500c431ba0ae9e2f56985f9990aa34fa42fe048a10d26e52b40f2df2.c with opts {"shell":"/usr/bin/bash"}
./CryptOpt: line 31: 4149251 Segmentation fault      PATH="$(realpath ./bins/node/bin):${PATH}" /usr/bin/env node "./dist/CryptOpt.js" "${@}"
```

gdb says:
```
Thread 1 "node" received signal SIGSEGV, Segmentation fault.
0x00007ffff7e25fd6 in ms_initialize () from /home/andreser/CryptOpt/node_modules/measuresuite/build/Release/measuresuite.node
(gdb) bt
#0  0x00007ffff7e25fd6 in ms_initialize ()
   from /home/andreser/CryptOpt/node_modules/measuresuite/build/Release/measuresuite.node
#1  0x00007ffff7e25b1b in init ()
   from /home/andreser/CryptOpt/node_modules/measuresuite/build/Release/measuresuite.node
```

<details>
<summary>more backtrace</summary>
<code>
#2  0x0000000000b10d7d in v8impl::(anonymous namespace)::FunctionCallbackWrapper::Invoke(v8::FunctionCallbackInfo<v8::Value> const&) ()
#3  0x0000000000db0230 in v8::internal::MaybeHandle<v8::internal::Object> v8::internal::(anonymous namespace)::HandleApiCallHelper<false>(v8::internal::Isolate*, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::FunctionTemplateInfo>, v8::internal::Handle<v8::internal::Object>, v8::internal::BuiltinArguments) ()
#4  0x0000000000db176f in v8::internal::Builtin_HandleApiCall(int, unsigned long*, v8::internal::Isolate*) ()
#5  0x00000000016ef579 in Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_BuiltinExit ()
#6  0x00000000016734d0 in Builtins_InterpreterEntryTrampoline ()
#7  0x00002e0bd55015b9 in ?? ()
#8  0x000015cb6c393811 in ?? ()
#9  0x0000000800000000 in ?? ()
#10 0x00002e0bd5501689 in ?? ()
#11 0x00003ebecb2fef29 in ?? ()
#12 0x0000000500000000 in ?? ()
#13 0x0000000100000000 in ?? ()
#14 0x0000000100000000 in ?? ()
#15 0x0000000100000000 in ?? ()
#16 0x0000000100000000 in ?? ()
#17 0x0000000500000000 in ?? ()
#18 0x00003ebecb2fef29 in ?? ()
#19 0x000015cb6c393811 in ?? ()
#20 0x0000064757026351 in ?? ()
#21 0x0000009400000000 in ?? ()
#22 0x000027d6971beb29 in ?? ()
#23 0x0000000000000008 in ?? ()
#24 0x00003a9153c02699 in ?? ()
#25 0x0000064757026351 in ?? ()
#26 0x00007fffffff9220 in ?? ()
#27 0x0000000001670e62 in Builtins_JSConstructStubGeneric ()
Backtrace stopped: frame did not save the PC
</code>
</details>

<details>
<summary>
<code>(gdb) disas $pc</code>
</summary>
<code>
Dump of assembler code for function ms_initialize:
   0x00007ffff7e25f30 <+0>:     push   %rbp
   0x00007ffff7e25f31 <+1>:     xor    %eax,%eax
   0x00007ffff7e25f33 <+3>:     mov    %rsp,%rbp
   0x00007ffff7e25f36 <+6>:     push   %r15
   0x00007ffff7e25f38 <+8>:     mov    %ecx,%r15d
   0x00007ffff7e25f3b <+11>:    push   %r14
   0x00007ffff7e25f3d <+13>:    mov    %edx,%r14d
   0x00007ffff7e25f40 <+16>:    push   %r13
   0x00007ffff7e25f42 <+18>:    mov    %esi,%r13d
   0x00007ffff7e25f45 <+21>:    push   %r12
   0x00007ffff7e25f47 <+23>:    mov    %rdi,%r12
   0x00007ffff7e25f4a <+26>:    push   %rbx
   0x00007ffff7e25f4b <+27>:    sub    $0x8,%rsp
   0x00007ffff7e25f4f <+31>:    call   0x7ffff7e24400 <create_default_ms@plt>
   0x00007ffff7e25f54 <+36>:    mov    %rax,(%r12)
   0x00007ffff7e25f58 <+40>:    test   %rax,%rax
   0x00007ffff7e25f5b <+43>:    je     0x7ffff7e25f6f <ms_initialize+63>
   0x00007ffff7e25f5d <+45>:    mov    %r13d,%esi
   0x00007ffff7e25f60 <+48>:    mov    %rax,%rdi
   0x00007ffff7e25f63 <+51>:    mov    %rax,%rbx
   0x00007ffff7e25f66 <+54>:    call   0x7ffff7e24580 <set_argwidth@plt>
   0x00007ffff7e25f6b <+59>:    test   %eax,%eax
   0x00007ffff7e25f6d <+61>:    je     0x7ffff7e25f88 <ms_initialize+88>
   0x00007ffff7e25f6f <+63>:    mov    $0x1,%eax
   0x00007ffff7e25f74 <+68>:    add    $0x8,%rsp
   0x00007ffff7e25f78 <+72>:    pop    %rbx
   0x00007ffff7e25f79 <+73>:    pop    %r12
   0x00007ffff7e25f7b <+75>:    pop    %r13
   0x00007ffff7e25f7d <+77>:    pop    %r14
   0x00007ffff7e25f7f <+79>:    pop    %r15
   0x00007ffff7e25f81 <+81>:    pop    %rbp
   0x00007ffff7e25f82 <+82>:    ret
   0x00007ffff7e25f83 <+83>:    nopl   0x0(%rax,%rax,1)
   0x00007ffff7e25f88 <+88>:    mov    %r14d,%esi
   0x00007ffff7e25f8b <+91>:    mov    %rbx,%rdi
   0x00007ffff7e25f8e <+94>:    call   0x7ffff7e24540 <set_argin@plt>
   0x00007ffff7e25f93 <+99>:    test   %eax,%eax
   0x00007ffff7e25f95 <+101>:   jne    0x7ffff7e25f6f <ms_initialize+63>
   0x00007ffff7e25f97 <+103>:   mov    %r15d,%esi
   0x00007ffff7e25f9a <+106>:   mov    %rbx,%rdi
   0x00007ffff7e25f9d <+109>:   call   0x7ffff7e242b0 <set_argout@plt>
   0x00007ffff7e25fa2 <+114>:   test   %eax,%eax
   0x00007ffff7e25fa4 <+116>:   jne    0x7ffff7e25f6f <ms_initialize+63>
   0x00007ffff7e25fa6 <+118>:   mov    %rbx,%rdi
   0x00007ffff7e25fa9 <+121>:   call   0x7ffff7e24460 <validate_num_args@plt>
   0x00007ffff7e25fae <+126>:   test   %eax,%eax
   0x00007ffff7e25fb0 <+128>:   jne    0x7ffff7e25f6f <ms_initialize+63>
   0x00007ffff7e25fb2 <+130>:   mov    %rbx,%rdi
   0x00007ffff7e25fb5 <+133>:   call   0x7ffff7e241f0 <init_random@plt>
   0x00007ffff7e25fba <+138>:   test   %eax,%eax
   0x00007ffff7e25fbc <+140>:   jne    0x7ffff7e25f6f <ms_initialize+63>
   0x00007ffff7e25fbe <+142>:   mov    %rbx,%rdi
   0x00007ffff7e25fc1 <+145>:   call   0x7ffff7e246d0 <init_json@plt>
   0x00007ffff7e25fc6 <+150>:   test   %eax,%eax
   0x00007ffff7e25fc8 <+152>:   jne    0x7ffff7e25f6f <ms_initialize+63>
   0x00007ffff7e25fca <+154>:   mov    %rbx,%rdi
   0x00007ffff7e25fcd <+157>:   call   0x7ffff7e24560 <init_timer@plt>
   0x00007ffff7e25fd2 <+162>:   test   %eax,%eax
   0x00007ffff7e25fd4 <+164>:   jne    0x7ffff7e25f6f <ms_initialize+63>
=> 0x00007ffff7e25fd6 <+166>:   movl   $0x0,0x88(%rbx)
   0x00007ffff7e25fe0 <+176>:   jmp    0x7ffff7e25f74 <ms_initialize+68>
</code>
</details>

<details>
<summary>
(gdb) info registers
</summary>
<code>
rax            0x0                 0
rbx            0x80eefa3fe0        553765191648
rcx            0x7ffff7b20733      140737349027635
rdx            0xffffffffffffff00  -256
rsi            0x1000              4096
rdi            0x0                 0
rbp            0x7fffffff8e60      0x7fffffff8e60
rsp            0x7fffffff8e30      0x7fffffff8e30
r8             0xfffffff9          4294967289
r9             0x0                 0
r10            0x1                 1
r11            0x246               582
r12            0x7fffffff8e88      140737488326280
r13            0x5                 5
r14            0x1                 1
r15            0x1                 1
rip            0x7ffff7e25fd6      0x7ffff7e25fd6 <ms_initialize+166>
eflags         0x10246             [ PF ZF IF RF ]
cs             0x33                51
ss             0x2b                43
ds             0x0                 0
es             0x0                 0
fs             0x0                 0
gs             0x0                 0
</code>
</details>

<details>
<summary> node-segfault-handler log, read after the same gdb execution: </summary>
<code>
-----[ Native Stacktraces ]-----
[pc=0x00007ffff7fb569e, sp=0x00007fffffff8860] in segfault_handler(int)+0x4e
[pc=0x00007ffff7a5af90, sp=0x00007fffffff8880] in __sigaction+0x40
[pc=0x00007ffff7e25fd6, sp=0x00007fffffff8e30] in ms_initialize+0xa6
[pc=0x00007ffff7e25b1b, sp=0x00007fffffff8e70] in init+0xab
[pc=0x0000000000b10d7d, sp=0x00007fffffff8ed0] in v8impl::(anonymous namespace)::FunctionCallbackWrapper::Invoke(v8::FunctionCallbackInfo<v8::Value> const&)+0x7d
[pc=0x0000000000db0230, sp=0x00007fffffff8f30] in v8::internal::MaybeHandle<v8::internal::Object> v8::internal::(anonymous namespace)::HandleApiCallHelper<false>(v8::internal::Isolate*, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::FunctionTemplateInfo>, v8::internal::Handle<v8::internal::Object>, v8::internal::BuiltinArguments)+0x380
[pc=0x0000000000db176f, sp=0x00007fffffff9070] in v8::internal::Builtin_HandleApiCall(int, unsigned long*, v8::internal::Isolate*)+0xaf
[pc=0x00000000016ef579, sp=0x00007fffffff90e0] in Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_BuiltinExit+0x39
[pc=0x00000000016734d0, sp=0x00007fffffff90f0] in Builtins_InterpreterEntryTrampoline+0xd0
[pc=0x0000000001670e62, sp=0x00007fffffff9100] in Builtins_JSConstructStubGeneric+0x122
[pc=0x000000000178d3ff, sp=0x00007fffffff9110] in Builtins_ConstructHandler+0x2bf
[pc=0x00000000016734d0, sp=0x00007fffffff9120] in Builtins_InterpreterEntryTrampoline+0xd0
[pc=0x00000000016734d0, sp=0x00007fffffff9130] in Builtins_InterpreterEntryTrampoline+0xd0
[pc=0x00000000016734d0, sp=0x00007fffffff9140] in Builtins_InterpreterEntryTrampoline+0xd0
[pc=0x0000000001670e62, sp=0x00007fffffff9150] in Builtins_JSConstructStubGeneric+0x122
[pc=0x000000000178d3ff, sp=0x00007fffffff9160] in Builtins_ConstructHandler+0x2bf
[pc=0x00000000016734d0, sp=0x00007fffffff9170] in Builtins_InterpreterEntryTrampoline+0xd0
[pc=0x00000000016f96d7, sp=0x00007fffffff9180] in Builtins_AsyncModuleEvaluate+0x97
[pc=0x0000000001671adc, sp=0x00007fffffff9190] in Builtins_JSEntryTrampoline+0x5c
[pc=0x0000000001671803, sp=0x00007fffffff91a0] in Builtins_JSEntry+0x83
[pc=0x0000000000e9293e, sp=0x00007fffffff91b0] in v8::internal::(anonymous namespace)::Invoke(v8::internal::Isolate*, v8::internal::(anonymous namespace)::InvokeParams const&)+0x13e
[pc=0x0000000000e93a10, sp=0x00007fffffff9a50] in v8::internal::(anonymous namespace)::InvokeWithTryCatch(v8::internal::Isolate*, v8::internal::(anonymous namespace)::InvokeParams const&) [clone .constprop.84]+0x50
[pc=0x0000000000e94012, sp=0x00007fffffff9ab0] in v8::internal::Execution::TryCall(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, int, v8::internal::Handle<v8::internal::Object>*, v8::internal::Execution::MessageHandling, v8::internal::MaybeHandle<v8::internal::Object>*, bool)+0x62
[pc=0x00000000011c3611, sp=0x00007fffffff9b40] in v8::internal::SourceTextModule::InnerExecuteAsyncModule(v8::internal::Isolate*, v8::internal::Handle<v8::internal::SourceTextModule>, v8::internal::Handle<v8::internal::JSPromise>)+0xd1
[pc=0x00000000011c38df, sp=0x00007fffffff9ba0] in v8::internal::SourceTextModule::ExecuteAsyncModule(v8::internal::Isolate*, v8::internal::Handle<v8::internal::SourceTextModule>)+0x1af
[pc=0x00000000011c5785, sp=0x00007fffffff9c10] in v8::internal::SourceTextModule::AsyncModuleExecutionFulfilled(v8::internal::Isolate*, v8::internal::Handle<v8::internal::SourceTextModule>)+0x135
[pc=0x0000000000dbdc2b, sp=0x00007fffffff9ce0] in v8::internal::Builtin_CallAsyncModuleFulfilled(int, unsigned long*, v8::internal::Isolate*)+0x3b
[pc=0x00000000016ef579, sp=0x00007fffffff9d20] in Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_BuiltinExit+0x39
[pc=0x000000000173fa71, sp=0x00007fffffff9d30] in Builtins_PromiseFulfillReactionJob+0x31
[pc=0x000000000169893b, sp=0x00007fffffff9d40] in Builtins_RunMicrotasks+0x27b
[pc=0x0000000001671a03, sp=0x00007fffffff9d50] in Builtins_JSRunMicrotasksEntry+0x83
[pc=0x0000000000e92e0a, sp=0x00007fffffff9d60] in v8::internal::(anonymous namespace)::Invoke(v8::internal::Isolate*, v8::internal::(anonymous namespace)::InvokeParams const&)+0x60a
[pc=0x0000000000e93a10, sp=0x00007fffffff9fb0] in v8::internal::(anonymous namespace)::InvokeWithTryCatch(v8::internal::Isolate*, v8::internal::(anonymous namespace)::InvokeParams const&) [clone .constprop.84]+0x50
[pc=0x0000000000e9411a, sp=0x00007fffffffa010] in v8::internal::Execution::TryRunMicrotasks(v8::internal::Isolate*, v8::internal::MicrotaskQueue*, v8::internal::MaybeHandle<v8::internal::Object>*)+0x5a
[pc=0x0000000000ec161a, sp=0x00007fffffffa080] in v8::internal::MicrotaskQueue::RunMicrotasks(v8::internal::Isolate*) [clone .part.50]+0x8a
[pc=0x0000000000ec19f2, sp=0x00007fffffffa130] in v8::internal::MicrotaskQueue::PerformCheckpoint(v8::Isolate*)+0x42
[pc=0x0000000000aaec29, sp=0x00007fffffffa160] in node::InternalCallbackScope::Close()+0x109
[pc=0x0000000000aaee01, sp=0x00007fffffffa1b0] in node::InternalCallbackScope::~InternalCallbackScope()+0x11
[pc=0x0000000000b70e5b, sp=0x00007fffffffa1d0] in node::fs::FileHandle::CloseReq::Resolve()+0x9b
[pc=0x0000000000b73420, sp=0x00007fffffffa250] in node::fs::FileHandle::ClosePromise()::{lambda(uv_fs_s*)#1}::_FUN(uv_fs_s*)+0x230
[pc=0x0000000000b681f4, sp=0x00007fffffffa2d0] in node::MakeLibuvRequestCallback<uv_fs_s, void (*)(uv_fs_s*)>::Wrapper(uv_fs_s*)+0x54
[pc=0x000000000164e11d, sp=0x00007fffffffa300] in uv__work_done+0x9d
[pc=0x0000000001652906, sp=0x00007fffffffa350] in uv__async_io.part.1+0x126
[pc=0x0000000001664e44, sp=0x00007fffffffa7a0] in uv__io_poll+0x494
[pc=0x000000000165326e, sp=0x00007fffffffd8b0] in uv_run+0x14e
[pc=0x0000000000aafa2d, sp=0x00007fffffffd910] in node::SpinEventLoop(node::Environment*)+0x14d
[pc=0x0000000000bb11f4, sp=0x00007fffffffd9c0] in node::NodeMainInstance::Run()+0xf4
[pc=0x0000000000b26c44, sp=0x00007fffffffda50] in node::LoadSnapshotDataAndRun(node::SnapshotData const**, node::InitializationResult const*)+0xb4
[pc=0x0000000000b2a83f, sp=0x00007fffffffdb00] in node::Start(int, char**)+0x2df
[pc=0x00007ffff7a4618a, sp=0x00007fffffffdb70] in __libc_init_first+0x8a
[pc=0x00007ffff7a46245, sp=0x00007fffffffdc10] in __libc_start_main+0x85
[pc=0x0000000000aad7ee, sp=0x00007fffffffdc60] in _start+0x2e

---[ V8 JavaScript Stacktraces ]---
at Measuresuite (file:///home/andreser/CryptOpt/node_modules/measuresuite/ts/dist/index.js:62:16)
at createMS (file:///home/andreser/CryptOpt/dist/optimizer.helper.class-13535478.js:4108:23)
at init (file:///home/andreser/CryptOpt/dist/optimizer.helper.class-13535478.js:4129:12)
at Optimizer (file:///home/andreser/CryptOpt/dist/CryptOpt.js:4535:46)
at (null) (file:///home/andreser/CryptOpt/dist/CryptOpt.js:4856:20)
</code>
</details>

`$ node --version`
`v18.13.0`

**Comment by dderjoel** [comment]
Interesting. 
So it seems to be an issue with the meausuresuite submodule.
So I have seen Segfaults with that (but I was not able to reproduce and fix yet),  but not when I was using CryptOpt.
So this is new. It seems to error with `__sigaction` ...
I do remember that I've created a check in the measuresuite tests, which listens to `SIGILL`'s, which may occur if the cpu does not have BMI2 / ADX. But that I believe was only in the tests, not in the lib-code. see here:
https://github.com/0xADE1A1DE/MeasureSuite/blob/main/lib/test/helper.h#L65

So you run in a VM, container, WLS or on some special emulation something ? 
Does your CPU have the flags BMI2 / ADX?

what you could do, is cloning this repo `https://github.com/0xADE1A1DE/MeasureSuite` and run a `make check` in there lets see if that error really comes from there.

**Comment by andres-erbsen** [comment]
It's a VM on a proprietary KVM-based hypervisor. Yes, bmi2 and adx are detected. CryptOpt worked inside Docker on a throwaway vm without known differences. `make check` passes in AssemblyLine and MeasureSuite.

node-segfault handler probably listens for signals, but the segfault also occurs without it.

**Comment by andres-erbsen** [comment]
The following script reproduces a similar segfault in a clean chroot:

```sh
#!/bin/sh

# apt-get install cdebootstrap
mkdir /tmp/chroot-debian-testing
cdebootstrap sid /tmp/chroot-debian-testing http://deb.debian.org/debian/
mount proc /tmp/chroot-debian-testing/proc -t proc
mount sysfs /tmp/chroot-debian-testing/sys -t sysfs
chroot /tmp/chroot-debian-testing << EOF
  apt-get update
  apt-get install -y vim zsh git curl tar xz-utils make autoconf g++ pkg-config libtool python-is-python3 jq

  git clone --recursive --depth=1 https://github.com/0xADE1A1DE/AssemblyLine.git
  cd AssemblyLine/
    ./autogen.sh
    ./configure
    make CFLAGS=-O3 all install
    cd ..

  git clone --recursive --depth=1 https://github.com/0xADE1A1DE/CryptOpt.git
  cd CryptOpt/
    make
    make check
    cd ..
  exit
EOF

umount /tmp/chroot-debian-testing/proc
umount /tmp/chroot-debian-testing/sys
```

**Comment by dderjoel** [comment]
alright, I'll try to reproduce

**Comment by andres-erbsen** [comment]
Did you get a chance to try this out?

**Comment by andres-erbsen** [comment]
> :confused:

I meant have you checked whether https://github.com/0xADE1A1DE/CryptOpt/issues/118#issuecomment-1485958383 also results in a segfault on your machine. Is this something that you can easily do, or should I pursue another avenue for moving forward here?

**Comment by dderjoel** [comment]
I cannot reproduce.
Using your script from comment on my Ubuntu Server, does not work. It does extract some files, but then after 'P: Configuring package login' it errors with "E: Internal error install" without any more info.

Using this script on my archlinux, (replacing cdebootstrap with debootstrap), it errors with 
```
/usr/share/debootstrap/functions: line 1665: /tmp/chroot-debian-testing/test-dev-null: Permission denied
E: Cannot install into target '/tmp/chroot-debian-testing' mounted with noexec or nodev
```
So I tried in a Fresh VM with a Debian 11 image.
It would not work with the Fiat binaries, because of GLIBC version incompatibilities. But that is orthogonal. 
The Bitcoin-code bridge works, which uses the measuresuite internally, too.

Is the `sid` in the bootstrap command the version of the debain image? Should I try this version? Or can you try Debian 11?

**Comment by andres-erbsen** [comment]
Uh, sounds like you are having as much fun as I was. I did expect `cdebootstrap` binaries from different OS repos to be compatible, without good reason, just experience, but I guess they're not. Yes, `sid` is the version I used; `stable` should be an alternative but I am not asking you to try that now. I'm not sure what to make of some tests passing on Debian 11; some tests also pass for me.

:thinking:

**Comment by dderjoel** [comment]
> I'm not sure what to make of some tests passing on Debian 11; some tests also pass for me.

So some tests are passing on deb 11 and some are not? Can you share which ones are passing? Not all are using measuresuite.

**Comment by dderjoel** [comment]
Speaking of Docker, maybe you can try to create a Docker image, which I can try to reproduce in?

**Comment by dderjoel** [comment]
more importantly, what is the value of `/proc/sys/kernel/perf_event_paranoid` ? if it is greater than `1`, try to `echo 1 | sudo tee /proc/sys/kernel/perf_event_paranoid` and run CryptOpt again.

**Comment by dderjoel** [comment]
I have encountered strange segfaults and stack smashing through Measuresuite lately. I fixed what I could reproduce locally with a work around. The fix is available in `measuresuite@2.0.2-beta.1`. If you clone the 'dev' branch here, this new version should be installed. If you have time, please have a test if the segfaults persist.

**Comment by andres-erbsen** [comment]
Here's the latest segfault on the dev branch: https://gist.github.com/andres-erbsen/bf0e232c54c9dba1ad1c4bce69a13248 ; `/proc/sys/kernel/perf_event_paranoid` is 1.

The issue does not reproduce inside Docker; the following Dockerfile seems to work: https://gist.github.com/andres-erbsen/dd06a33d7c92fef821069659fc7657de But running `make check` without docker still fails, on the same machine right after:

```
 ↓ test/fiat-curves/curve25519_solinas-mul2.ts (1) [skipped]
 · test/index.ts (2)
 · test/bitcoin-curves/secp256k1-mul.ts (1)
 · test/bitcoin-curves/secp256k1-square.ts (1)
 ❯ test/fiat-curves/curve25519-mul.ts (1)
   · optimise
 ❯ test/fiat-curves/curve25519-square.ts (1)
   · optimise
 · test/fiat-curves/curve25519_solinas-mul.ts (1)
make: *** [Makefile:51: check] Segmentation fault
```

**Comment by dderjoel** [comment]
> more importantly, what is the value of `/proc/sys/kernel/perf_event_paranoid` ? if it is greater than `1`, try to `echo 1 | sudo tee /proc/sys/kernel/perf_event_paranoid` and run CryptOpt again.

Can you check that?

**Comment by dderjoel** [comment]
> > more importantly, what is the value of `/proc/sys/kernel/perf_event_paranoid` ? if it is greater than `1`, try to `echo 1 | sudo tee /proc/sys/kernel/perf_event_paranoid` and run CryptOpt again.
> 
> Can you check that?

never mind. I've Just re read your message. its 1.

Also very interesting that the bitcoin curves run through...

**Comment by dderjoel** [comment]
TLDR: I've updated measuresuite once again. available with the Cryptopt-dev branch.

Longer Story:
I've encountered very strange behaviour. If I run one instance for long (e.g. with the `--single` parameter) it works fine, but if I use many bets it segfaults at some point in the `timer.c`. The difference in those two modes is that the object `Measuresuite` will get allocated on the heap with every "bet".
I've changed two things: 
- explicitly free memory (`munmap` in `destroy`-functions) when the optimizer is done with one bet.
-  fix a (severe) bug in `timer.c` (How did that ever work)? The number for the Systemcall to get the page for the PMC was written into `a` but it should be in `rdi`. This is done at initialization phase, so your error could actually be fixed.

**Comment by andres-erbsen** [comment]
This seems to have fixed it. Thank you!

```
 Test Files  58 passed | 1 skipped (59)
      Tests  806 passed | 2 skipped (808)
   Start at  15:09:21
   Duration  40.30s (transform 32.30s, setup 27ms, collect 853.09s, tests 144.48s, environment 89ms, prepare 40.61s)
```

I also see `mmap (for performance counter) failed: Bad file descriptor` (several times) even though `andreser@andreser ~/CryptOpt % cat /proc/sys/kernel/perf_event_paranoid` returns `1`.

**Comment by dderjoel** [comment]
Hm, okay, that may happen, I'm unsure about the reason, but MS gracefully falls back to use `rdtscp`
